mithril-aggregator/src/services/prover.rs

```diff
@@ -70,25 +70,37 @@ impl MithrilProverService {
         &self,
         transactions: Vec<CardanoTransaction>,
     ) -> StdResult<MKMap<BlockRange, MKMapNode<BlockRange>>> {
-        let mut transactions_by_block_ranges: HashMap<BlockRange, Vec<TransactionHash>> =
-            HashMap::new();
+        let mut transactions_by_block_ranges: BTreeMap<BlockRange, Vec<TransactionHash>> =
+            BTreeMap::new();
+        let mut last_transaction: Option<CardanoTransaction> = None;
         for transaction in transactions {
             let block_range = BlockRange::from_block_number(transaction.block_number);
+            last_transaction = Some(transaction.clone());
             transactions_by_block_ranges
                 .entry(block_range)
                 .or_default()
                 .push(transaction.transaction_hash);
         }
+        let mut block_ranges = transactions_by_block_ranges.into_iter().try_fold(
+            vec![],
+            |mut acc, (block_range, transactions)| -> StdResult<Vec<(_, MKMapNode<_>)>> {
+                acc.push((block_range, MKTree::new(&transactions)?.into()));
+                Ok(acc)
+            },
+        )?;
+
+        // This is a temporary fix to avoid including an incomplete block ranges in the computation of the prover.
+        // This will be swiftly replaced by the new way of computing proof relying on the block range roots stored in database.
+        if let Some(transaction) = last_transaction {
+            if let Some((last_block_range, _)) = block_ranges.last() {
+                if transaction.block_number < last_block_range.end - 1 {
```

Comment:
Are you sure about this condition ? Is it verify by a test ? 
BlockRange.end is the last block of the BlockRange, so it will be " < last_block_range.end" or '<= last_block_range.end - 1"

Reply:
I guess it is good even though this is only tested with the e2e test:
- We want to discard transactions from an incomplete block range (e.g. **15..30**) when we prove transaction in order to mimic what's done on the signature side 
- If the last transaction is not in the last block of the last block range, we need to discard the block range
- If the last transaction is done e.g. on block **29** which is the last block of the block range: the block range is complete and we can keep it (`29 >= 30 -1`)
- If the last transaction is done e.g. on block **28** which is not the last block of the block range: the block range is incomplete and we can discard it (`28 < 30 - 1`)

Moreover this function will be deleted in the implementation of [#1635](https://github.com/input-output-hk/mithril/issues/1635).